### PR TITLE
AIP-45 Remove dag parsing in airflow run local

### PR DESCRIPTION
**Pull request by pingzh**
[AIP Link](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-45+Remove+double+dag+parsing+in+airflow+run)

Changes in this PR:

**1. Remove dag file parsing in `airflow tasks run --local` process**

1. Read serialized dag in `airflow tasks run --local` process
2. Add dag file parsing to `fork` in the `TaskRunner` since `airflow tasks run --local` process won't parse the dag_file

**2. Move callbacks from `airflow tasks run --local` process to `airflow tasks run --raw` process.**

1. After the AIP 45, the `airflow tasks run --local` process won't parse the dag file,  thus callbacks cannot be run in the `airflow task run --local` process. 
4.  Running callbacks in the `airflow tasks run --local` is not secure, when there is run_as_user for the task, the process owner of the `airflow tasks run --local` is `sudoable`, so the callbacks of the task are run by the sudoable user not the task's run_as_user.

---

**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by pingzh** [comment]
hi @potiuk , the AIP-45 is ready to review. could you please take a look?

**Comment by potiuk** [comment]
> hi @potiuk , the AIP-45 is ready to review. could you please take a look?

will do shortly

File: airflow/jobs/local_task_job.py [hunk 1]
```diff
@@ -104,11 +104,6 @@ def signal_handler(signum, frame):
         try:
             self.task_runner.start()
 
-            # Unmap the task _after_ it has forked/execed. (This is a bit of a kludge, but if we unmap before
```
Comment (potiuk) [inline airflow/jobs/local_task_job.py]: Unless I am mistaken, I believe this was removed by mistake (and the next piece of code as well). Those two (unmap task and handling signals have been added recently so likely ths is caused by starting off from previous version and not realising the code is removed when rebasing.

File: airflow/models/taskinstance.py [hunk 2]
```diff
@@ -1410,18 +1410,7 @@ def _execute_task_with_callbacks(self, context, test_mode=False):
         """Prepare Task for Execution"""
         from airflow.models.renderedtifields import RenderedTaskInstanceFields
 
-        parent_pid = os.getpid()
```
Comment (potiuk) [inline airflow/models/taskinstance.py]: Same here - this code is deleted by mistake I believe

File: docker_tests/test_docker_compose_quick_start.py [hunk 3]
```diff
@@ -102,8 +102,8 @@ def wait_for_container(container_id: str, timeout: int = 300):
 
 
 def wait_for_terminal_dag_state(dag_id, dag_run_id):
-    # Wait 30 seconds
-    for _ in range(30):
```
Comment (potiuk) [inline docker_tests/test_docker_compose_quick_start.py]: Can you make a separate PR with that change please. Seems unrelated?

**Review by potiuk — DISMISSED** [review]

**Comment by potiuk** [comment]
Hey @pingzh - can you address the comments and split this one into three PRs? The way we do it usually is that we have first PR that is with the first commit only and then second PR where we ask in the PR to only review the last commit. And there is a separate change with "increasing CI timeout" which also deserves to be separated. 

This makes it easier to review and merge those changes later independently or one-after-another.

File: docker_tests/test_docker_compose_quick_start.py (hunk 3 above)
Comment (pingzh) [reply docker_tests/test_docker_compose_quick_start.py]: good point, https://github.com/apache/airflow/pull/22476 thanks

File: airflow/jobs/local_task_job.py (hunk 1 above)
Comment (pingzh) [reply airflow/jobs/local_task_job.py]: @potiuk actually they were removed since the `task` in the `task_instance` is not longer needed as the LocalTaskJob not longer runs the callbacks and the `task` is deserialized from the database, calling `unmap` is unnecessary and also causes issues when the task involves `MappedOperator`, https://github.com/apache/airflow/blob/main/airflow/models/mappedoperator.py#L455-L456 

```
        if isinstance(self.operator_class, str):
            raise RuntimeError("Cannot unmap a deserialized operator")
```

(the `operator_class` is str for deserialized object)

File: airflow/models/taskinstance.py (hunk 2 above)
Comment (pingzh) [reply airflow/models/taskinstance.py]: @potiuk this check is no needed since the `TaskInstance#_run_raw_task` is always in separated process as this PR removes the process fork from the `LocalTaskJob`. Forking from the `LocalTaskJob` might have security issues when there is run_as_user for the task.

Comment (potiuk) [reply airflow/models/taskinstance.py]: That's why it needs to be separated to a separate PR where you explain your intentions in detail in the PR description. This information was missing in the original commit "Remove dag parsing in airflow run" did not explain any of the context here.

**Comment by pingzh** [comment]
> Hey @pingzh - can you address the comments and split this one into three PRs? The way we do it usually is that we have first PR that is with the first commit only and then second PR where we ask in the PR to only review the last commit. And there is a separate change with "increasing CI timeout" which also deserves to be separated.
> 
> This makes it easier to review and merge those changes later independently or one-after-another.

@potiuk thanks a lot for reviewing the PR. i have rebased the main branch with bumping the CI timeout change.

Good point, I have separated them into 2 PRs however we need to merge the second commit before the first commit.

File: airflow/models/taskinstance.py (hunk 2 above)
Comment (potiuk) [reply airflow/models/taskinstance.py]: BTW. I think we should not remove forking. If you want to remove it , then this should be a separate PR where you explain why. Forking is shaving multiple seconds when startting the task sometimes because it does not have to launch the interpreter. I'd even say that we won't remove forking because of that. There shuld be very good reason why we would like to remove it and some justification on how it impacts the task start time.

Comment (pingzh) [reply airflow/models/taskinstance.py]: @potiuk actually on a second thoughts, these 2 commits are intertwined. How about let me merge those 2 commits into single one and add more description to the commit message and the PR description. 

let me know your thoughts.

Comment (pingzh) [reply airflow/models/taskinstance.py]: > BTW. I think we should not remove forking. If you want to remove it , then this should be a separate PR where you explain why. Forking is shaving multiple seconds when startting the task sometimes because it does not have to launch the interpreter. I'd even say that we won't remove forking because of that. There shuld be very good reason why we would like to remove it and some justification on how it impacts the task start time.

updated the pr description to reflect the reason for removing `fork` in the `TaskRunner`

File: airflow/jobs/local_task_job.py (hunk 1 above)
Comment (potiuk) [reply airflow/jobs/local_task_job.py]: Welll. I think this is not really good idea to remove the fork. Correct me if I am wrong, but this will ALWAYS add overhead for starting the process and launching a new interpreter, where `fork` was supposed to speed it up immensely. And IMHO it has nothing to do with where DAGs are parsed.

While it might not matter for your case where you have huge DAGs to parse and "fork" vs. "launch process" overhead is negligible, it will significantly change the characteristics and overhead in case the DAGs are small (because the whole interpreter will need to be launched in order to run the task. This impacts both memory and time and I think we should not remove this optimization.

With fork, we use the fact that forking costs almost nothing - both in terms of time and memory. Forked processes use copy-on-write semantics for memory used, which means that most of the dynamic memory allocated before forking will not be re-allocated again in the fork (unless the memory changes).

Unless I missed something - this is a huge change and we should not remove fork option.

Comment (potiuk) [reply airflow/jobs/local_task_job.py]: cc: @ashb @uranusjr @malthe  -> the usuall summoning here for your comments.

Comment (potiuk) [reply airflow/jobs/local_task_job.py]: BTW. my comment is slightly misplaced, but it concerns the whole PR and removing fork so I will leave it here (but it relates to a number of changes in this PR that remove "fork" option when starting task.

Comment (uranusjr) [reply airflow/jobs/local_task_job.py]: FWIW I just finished implementing (as in a few hours ago) unmapping a deserialised operator in #22396 so this may be irralevant now.

Comment (potiuk) [reply airflow/jobs/local_task_job.py]: > FWIW I just finished implementing (as in a few hours ago) unmapping a deserialised operator in #22396 so this may be irralevant now.

Yeah. I think this is more relevant to removal of the forking in general :) (misplaced comment sorry).

Comment (pingzh) [reply airflow/jobs/local_task_job.py]: @potiuk i have updated the pr description to reflect why the `fork` needs to be removed.  

1. `airflow tasks run --local` process no longer parses dag files. So forking this process won't let the `airflow tasks run --raw` get the parsed dag object,  unless we add additional logic to parse dag file in the fork. 
2. The TaskRunner only forks the `airflow tasks run --local` process when there is no run_as_user. 

I agree that fork can speed up `airflow tasks run --raw` process. I can add it back in this PR so that we can focus mainly on the main topic of AIP 45

**Comment by pingzh** [comment]
@potiuk i have added back the `fork` support in this PR. could you please take a look? thanks

File: airflow/jobs/scheduler_job.py [hunk 4]
```diff
@@ -779,6 +779,22 @@ def _execute(self) -> None:
                 self.log.exception("Exception when executing DagFileProcessorAgent.end")
             self.log.info("Exited execute loop")
 
+    def _update_dag_run_state_for_paused_dags(self):
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py:768]: Why did this have to move out of LocalTaskJob?

File: airflow/models/serialized_dag.py [hunk 5]
```diff
@@ -250,6 +250,13 @@ def has_dag(cls, dag_id: str, session: Session = None) -> bool:
         """
         return session.query(literal(True)).filter(cls.dag_id == dag_id).first() is not None
 
+    @classmethod
+    def get_dag(cls, dag_id: str) -> Optional['SerializedDAG']:
```
Comment (ashb) [inline airflow/models/serialized_dag.py]: This should take a session argument

File: airflow/models/taskinstance.py [hunk 6]
```diff
@@ -710,6 +670,26 @@ def mark_success_url(self):
             "&state=success"
         )
 
+    def current_pid(self) -> str:
```
Comment (ashb) [inline airflow/models/taskinstance.py]: This needs to look at hostname too otherwise the pid on different hosts could clash

File: airflow/jobs/scheduler_job.py (hunk 4 above)
Comment (pingzh) [reply airflow/jobs/scheduler_job.py:768]: since no callbacks is executed in the LocalTaskJob, we need to rely on the scheduler to check and update the paused dag runs.

File: airflow/models/serialized_dag.py (hunk 5 above)
Comment (pingzh) [reply airflow/models/serialized_dag.py]: good call. let me add it

File: airflow/models/taskinstance.py (hunk 6 above)
Comment (pingzh) [reply airflow/models/taskinstance.py]: this is a helper method to get the `pid` attached to the ti, just like `current_state`

Comment (ashb) [reply airflow/models/taskinstance.py]: It doesn't appear to be used anywhere?

Comment (pingzh) [reply airflow/models/taskinstance.py]: ah, good call. i must have used it but then forgot to delete it. let me remove it. thanks

File: airflow/jobs/scheduler_job.py [hunk 7]
```diff
@@ -783,6 +783,22 @@ def _execute(self) -> None:
                     self.log.exception("Exception when executing DagFileProcessorAgent.end")
             self.log.info("Exited execute loop")
 
+    def _update_dag_run_state_for_paused_dags(self):
+        try:
+            paused_dag_ids = DagModel.get_all_paused_dag_ids()
+            for dag_id in paused_dag_ids:
```
Comment (ephraimbuddy) [inline airflow/jobs/scheduler_job.py:771]: Hmm, for a deployment with 3000 paused dags, I wonder if there would be a challenge here? (probably unimaginable to pause all that..)

File: airflow/utils/cli.py [hunk 8]
```diff
@@ -213,6 +213,16 @@ def get_dag(subdir: Optional[str], dag_id: str) -> "DAG":
     return dagbag.dags[dag_id]
 
 
+def get_dag_by_deserialization(dag_id: str) -> "DAG":
+    from airflow.models.serialized_dag import SerializedDagModel
+
+    dag_model = SerializedDagModel.get(dag_id)
+    if dag_model is None:
+        raise AirflowException(f"Serialized dag_id could not be found: {dag_id}")
```
Comment (ephraimbuddy) [inline airflow/utils/cli.py]:
```suggestion
        raise AirflowException(f"Serialized DAG: {dag_id} could not be found")
```
How about this?

File: tests/jobs/test_local_task_job.py [hunk 9]
```diff
@@ -276,39 +278,26 @@ def test_heartbeat_failed_fast(self):
...
-            session.merge(ti)
-            session.commit()
-            # The below code will not run as heartbeat will detect change of state
-            time.sleep(10)
-
-        with dag_maker('test_mark_success'):
-            task = PythonOperator(task_id="task1", python_callable=task_function)
-        dr = dag_maker.create_dagrun()
+        dag = get_test_dag('test_mark_state')
+        with create_session() as session:
```
Comment (ephraimbuddy) [inline tests/jobs/test_local_task_job.py]: We have `session` fixture

File: airflow/jobs/scheduler_job.py (hunk 4 above)
Comment (ashb) [reply airflow/jobs/scheduler_job.py:768]: I'm not too happy about this change -- the point of having this in LocalTaskJob is so that the state of a dag run updates quickly after the task finishes.

By moving it to the scheduler we've lost that.

Instead of moving this to the scheduler, could we move it "down" in to the raw task process instead? That or I wonder if there is some way with AIP-44 to add a callback from the LTJ if needed.

Final option: _If_ there are no callbacks we can still safely do this here, so maybe we only need to do this in the scheduler in case of callbacks?

File: airflow/utils/cli.py (hunk 8 above)
Comment (pingzh) [reply airflow/utils/cli.py]: 👍

File: airflow/jobs/scheduler_job.py (hunk 4 above)
Comment (pingzh) [reply airflow/jobs/scheduler_job.py:768]: good point on having the LocalTaskJob to mark the dag run quickly. my concern is that it leaks the responsibility of the scheduler, also there is chance that the LTJ fails before marking the dag run state, which leaves the unhandled dag run states (this could be rare). also, since the dag is paused, it would be a good deal if the dag run state is updated a little bit late.

as we discussed in this email thread: `[DISCUSSION] let scheduler heal tasks stuck in queued state`, we will need to define the responsibility of each component in airflow (scheduler, executor, LTJ (airflow run --local), `airflow run --raw`) in terms of the state machine. 

we can have a thorough discussion there. let me know your thoughts.

File: airflow/jobs/scheduler_job.py (hunk 7 above)
Comment (pingzh) [reply airflow/jobs/scheduler_job.py:771]: good call. how about we adding a caching. if a paused dag_id is checked, we don't need to check it every time.

Comment (ephraimbuddy) [reply airflow/jobs/scheduler_job.py:771]: It's possible that at the first check of the `dag_run`, the `dag_run` state could not be updated because the task instances are still running. If we cache the dag, and the task instances of the `dag_run` `complete` later on, we won't be able to update the `dagrun` state, right?

Comment (pingzh) [reply airflow/jobs/scheduler_job.py:771]: it is possible. we can put the dag_id to the cache (or visited) when there is no `dag_runs` with `running` state for that dag_id. In this case, we skip them. let me know your thoughts. thanks

File: tests/jobs/test_local_task_job.py (hunk 9 above)
Comment (pingzh) [reply tests/jobs/test_local_task_job.py]: good call. thanks. updated.

**Comment by pingzh** [comment]
@potiuk could you please take a look? thanks

**Review by KevinYang21 — APPROVED** [review]

**Comment by KevinYang21** [comment]
@potiuk Hi Jarek, it seems like your previous comments have been addressed. Do you mind take another look and see if we can dismiss the change request please? Thanks.

**Comment by KevinYang21** [comment]
Thanks everyone for help reviewing the PR. Given the PR has been open for a while and all comments have been addressed, if no more comments, I'll merge this PR tonight.

**Comment by potiuk** [comment]
same here.

**Comment by KevinYang21** [comment]
> same here.

Oops, saw this comment before seeing [the comment from the other](https://github.com/apache/airflow/pull/22917#issuecomment-1124391783) and didn't get what it means.. We can revert this if we think major changes are required or fix forward if minor.

File: airflow/jobs/scheduler_job.py (hunk 4 above)
Comment (potiuk) [reply airflow/jobs/scheduler_job.py:768]: I quite agree with no need to update pasused tasks state immediately - this makes this part of the changfe far more appealing. Since this change is merged - @ashb - if you stil have some concerns, we can continue discussing here and maybe make a follow-up change in case we think of a scenario where it might be problematic.

**Comment by potiuk** [comment]
No problem @KevinYang21 - Agree. 

One thing here, since this is quite "sensitive" change (this is the main reason why we have not merged it before 2.3.0 and had no time to dive deep as there is still some focus on getting 2.3.1 out - fixing some "blocker" teething problems with 2.3.0 - is this possible (or maybe you have already done so) to run this change in AirBnB staging environment with some heavy'ier load (I guess you are also planning to migrate to 2.3+ at some point, so maybe the main version of airflow after those few changes ported from your earlier "fork" (as I understand) is already good enough to be tested at bigger scale at AirBnB to catch  potential races etc. ? 

I think that would alleviate a number of concerns about the change here - i know for a fact those kind of "deep core" changes are heavily tested by Ash and team in the Astro test/qa/staging environments rather than only via unit tests, so having some "thumbsup" from AriBnb using the main version of Airflow with those changes included at some point in time would be a great thing :D

**Comment by potiuk** [comment]
I honestly think that would be a win-win for all involved parties if the main version of Airllow is more heavily tested at AirBnB - before release, so that you also can have greater confidence on migrating faster :D

**Comment by KevinYang21** [comment]
@potiuk I agree it would be a win-win.

We are already working on staging+stress testing this change with 2.2.4 release. We will update the results here. At the same time, the exact same logic included in this change, just on 1.10.4 release, has been running in our production for 1 year. Yes 1.10.4 is different than main but it should still be able to provide a decent level of confidence given it's the same logic running across the two versions.

**Comment by ashb** [comment]
What kind of performance difference does this make on 2.2.x btw? (Just curious, it still feels like the right thing to do)

**Comment by pingzh** [comment]
> What kind of performance difference does this make on 2.2.x btw? (Just curious, it still feels like the right thing to do)

@ashb I am still in the process of porting over this change to 2.2.4. 

Based on our current production based on 1.10.4, it helped to bring down average task start time from 24 seconds to 3 seconds (across all three airflow clusters) and reduce peak memory usage by around 45%. (some context here, all of our dags have `run_as_user` for security, thus the `fork` in StandardTaskRunner won't help).

**Comment by ashb** [comment]
I've just noticed that this causes a problem for the follow-on mini scheduler for mapped tasks. I guess that code path wasn't sufficiently unit tested.

DAG

```
import csv
import io
import os
import json
from datetime import datetime

from airflow import DAG
from airflow.decorators import task
from airflow.models.xcom_arg import XComArg
from airflow.providers.amazon.aws.hooks.s3 import S3Hook
from airflow.providers.amazon.aws.operators.s3 import S3ListOperator


with DAG(dag_id='mapped_s3', start_date=datetime(2022, 5, 19)) as dag:
    files = S3ListOperator(
        task_id="get_inputs",
        bucket="airflow-summit-2022",
        prefix="data_provider_a/{{ data_interval_end | ds }}/",
        delimiter='/',
        do_xcom_push=True,
    )

    @task
    def csv_to_json(aws_conn_id, input_bucket, key, output_bucket):
        hook = S3Hook(aws_conn_id=aws_conn_id)

        csv_data = hook.read_key(key, input_bucket)
        reader = csv.DictReader(io.StringIO(csv_data))

        output = io.BytesIO()

        for row in reader:
            output.write(json.dumps(row, indent=None).encode('utf-8'))
            output.write(b"\n")

        output.seek(0, os.SEEK_SET)
        hook.load_file_obj(output, key=key.replace(".csv", ".json"), bucket_name=output_bucket)

    csv_to_json.partial(
        aws_conn_id="aws_default", input_bucket=files.bucket, output_bucket="airflow-summit-2022-processed"
    ).expand(key=XComArg(files))
```

Error:

```
  File "/home/ash/code/airflow/airflow/airflow/jobs/local_task_job.py", line 253, in _run_mini_scheduler_on_child_tasks
    info = dag_run.task_instance_scheduling_decisions(session)
  File "/home/ash/code/airflow/airflow/airflow/utils/session.py", line 68, in wrapper
    return func(*args, **kwargs)
  File "/home/ash/code/airflow/airflow/airflow/models/dagrun.py", line 658, in task_instance_scheduling_decisions
    schedulable_tis, changed_tis, expansion_happened = self._get_ready_tis(
  File "/home/ash/code/airflow/airflow/airflow/models/dagrun.py", line 714, in _get_ready_tis
    expanded_tis, _ = schedulable.task.expand_mapped_task(self.run_id, session=session)
  File "/home/ash/code/airflow/airflow/airflow/models/mappedoperator.py", line 609, in expand_mapped_task
    operator.mul, self._resolve_map_lengths(run_id, session=session).values()
  File "/home/ash/code/airflow/airflow/airflow/models/mappedoperator.py", line 591, in _resolve_map_lengths
    expansion_kwargs = self._get_expansion_kwargs()
  File "/home/ash/code/airflow/airflow/airflow/models/mappedoperator.py", line 526, in _get_expansion_kwargs
    return getattr(self, self._expansion_kwargs_attr)
AttributeError: 'MappedOperator' object has no attribute 'mapped_op_kwargs'
```

**Comment by potiuk** [comment]
So this is great we gave it a bit time to mature in "main" :)

File: airflow/jobs/scheduler_job.py [hunk 10]
```diff
@@ -764,6 +765,26 @@ def _execute(self) -> None:
...
+                if dag_id in self._paused_dag_without_running_dagruns:
+                    continue
```
Comment (tanelk) [inline airflow/jobs/scheduler_job.py:773]: dag_ids are never removed from this set. If I unpause the DAG after this has run and then later re-pause it, then this method skips it.

Comment (pingzh) [reply airflow/jobs/scheduler_job.py:773]: good catch. let me open a PR.

File: airflow/task/task_runner/base_task_runner.py [hunk 11]
```diff
@@ -95,24 +89,16 @@ def __init__(self, local_task_job):
...
-            popen_prepend
-            + self._task_instance.command_as_list(
-                raw=True,
-                pickle_id=local_task_job.pickle_id,
-                mark_success=local_task_job.mark_success,
-                job_id=local_task_job.id,
-                pool=local_task_job.pool,
-                cfg_path=cfg_path,
-            )
-            + ["--error-file", self._error_file.name]
```
Comment (alicj) [inline airflow/task/task_runner/base_task_runner.py:108]: Hi @pingzh , thanks for working on Airflow! I'm curious to know why the `--error-file` parameter is removed here?
